sw/host/opentitanlib/src/bootstrap/legacy.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use mundane::hash::{Digest, Hasher, Sha256};
+use std::time::Duration;
+use zerocopy::AsBytes;
+
+use crate::bootstrap::{BootstrapOptions, UpdateProtocol};
+use crate::io::spi::{Target, Transfer};
+
+#[derive(AsBytes, Debug, Default)]
+#[repr(C)]
+struct FrameHeader {
+    hash: [u8; Frame::HASH_LEN],
+    frame_num: u32,
+    flash_offset: u32,
+}
+
+#[derive(AsBytes, Debug)]
+#[repr(C)]
+struct Frame {
+    header: FrameHeader,
+    data: [u8; Frame::DATA_LEN],
+}
+
+impl Default for Frame {
+    fn default() -> Self {
+        Frame {
+            header: Default::default(),
+            data: [0xff; Frame::DATA_LEN],
+        }
+    }
+}
+
+impl Frame {
+    const EOF: u32 = 0x8000_0000;
+    const DATA_LEN: usize = 2048 - std::mem::size_of::<FrameHeader>();
+    const HASH_LEN: usize = 32;
+
+    /// Computes the hash in the header.
+    fn header_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let frame = self.as_bytes();
+        let sha = Sha256::hash(&frame[Frame::HASH_LEN..]);
+        sha.bytes()
+    }
+
+    /// Computes the hash over the entire frame.
+    fn frame_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let sha = Sha256::hash(self.as_bytes());
+        let mut digest = sha.bytes();
+        // Touch up zeroes into ones.
+        for i in 0..digest.len() {
+            if digest[i] == 0 {
+                digest[i] = 1;
+            }
+        }
+        digest
+    }
+
+    /// Creates a sequence of frames based on a `payload` binary.
+    fn from_payload(payload: &[u8]) -> Vec<Frame> {
+        let mut frames = Vec::new();
+
+        let mut max_addr = (payload.len() + 3) / 4;
```

Comment:
You are right, it was confusing.  I have changed the convention of all the _addr variables so that they all count bytes, eliminating the multiplications by 4.